File: runtime/secure_openapi.go
package apptheory

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"reflect"
	"sort"
	"strings"
)

// OpenAPIAuthSchemes binds secure posture classes to document-level scheme names.
type OpenAPIAuthSchemes struct {
	Authenticated []string `json:"authenticated"`
	InternalOnly  []string `json:"internal_only"`
}

// SecureOpenAPISpec describes an exact HTTP projection for a SecureApp.
type SecureOpenAPISpec struct {
	Title           string                    `json:"title"`
	Version         string                    `json:"version"`
	Routes          []OpenAPIRouteSpec        `json:"routes"`
	SecuritySchemes map[string]map[string]any `json:"security_schemes"`
	AuthSchemes     OpenAPIAuthSchemes        `json:"auth_schemes"`
}

type secureOpenAPIJoin struct {
	route       SecureRoute
	description OpenAPIRouteSpec
	emittedPath string
	proxy       bool
	proxyName   string
}

// GenerateOpenAPI returns the deterministic secure-v1 OpenAPI document bound to this app's routes.
func (a *SecureApp) GenerateOpenAPI(spec SecureOpenAPISpec) (map[string]any, error) {
	if a == nil || a.core == nil {
		return nil, errors.New("apptheory: secure app is nil")
	}
	registered, err := secureHTTPRouteSet(a.Routes())
	if err != nil {
		return nil, err
	}
	described, order, err := secureOpenAPIDescriptions(spec.Routes)
	if err != nil {
		return nil, err
	}
	if validationErr := secureValidateExactRouteSet(registered, described); validationErr != nil {
		return nil, validationErr
	}
	schemes, authenticatedSchemes, internalSchemes, err := secureOpenAPISchemes(spec)
	if err != nil {
		return nil, err
	}
	joins, legacyRoutes, err := secureOpenAPIJoins(registered, described, order, authenticatedSchemes, internalSchemes)
	if err != nil {
		return nil, err
	}
	doc, err := GenerateOpenAPI(OpenAPISpec{Title: spec.Title, Version: spec.Version, Routes: legacyRoutes})
	if err != nil {
		return nil, err
	}
	if err := secureDecorateOpenAPI(doc, schemes, joins, authenticatedSchemes, internalSchemes); err != nil {
		return nil, err
	}
	return doc, nil
}

func secureHTTPRouteSet(routes []SecureRoute) (map[string]SecureRoute, error) {
	registered := map[string]SecureRoute{}
	for _, route := range routes {
		if route.Surface != SecureRouteHTTP {
			continue
		}
		key, _, err := canonicalRouteKey(route.Method, route.Path)
		if err != nil {
			return nil, err
		}
		registered[key] = route
	}
	return registered, nil
}

func secureOpenAPIDescriptions(routes []OpenAPIRouteSpec) (map[string]OpenAPIRouteSpec, []string, error) {
	described := map[string]OpenAPIRouteSpec{}
	order := make([]string, 0, len(routes))
	for _, description := range routes {
		key, path, err := canonicalRouteKey(description.Method, description.Path)
		if err != nil {
			return nil, nil, fmt.Errorf("apptheory: secure openapi route: %w", err)
		}
		if _, exists := described[key]; exists {
			return nil, nil, fmt.Errorf("apptheory: secure openapi route %s is duplicated", key)
		}
		description.Method = strings.TrimSpace(strings.ToUpper(description.Method))
		description.Path = path
		described[key] = description
		order = append(order, key)
	}
	return described, order, nil
}

func secureValidateExactRouteSet(registered map[string]SecureRoute, described map[string]OpenAPIRouteSpec) error {
	registeredKeys := make([]string, 0, len(registered))
	for key := range registered {
		registeredKeys = append(registeredKeys, key)
	}
	sort.Strings(registeredKeys)
	for _, key := range registeredKeys {
		if _, ok := described[key]; !ok {
			return fmt.Errorf("apptheory: secure openapi missing route %s", key)
		}
	}
	describedKeys := make([]string, 0, len(described))
	for key := range described {
		describedKeys = append(describedKeys, key)
	}
	sort.Strings(describedKeys)
	for _, key := range describedKeys {
		if _, ok := registered[key]; !ok {
			return fmt.Errorf("apptheory: secure openapi extra route %s", key)
		}
	}
	return nil
}

func secureOpenAPISchemes(spec SecureOpenAPISpec) (map[string]any, []string, []string, error) {
	schemes, err := secureCanonicalSecuritySchemes(spec.SecuritySchemes)
	if err != nil {
		return nil, nil, nil, err
	}
	authenticatedSchemes := normalizeScopeList(spec.AuthSchemes.Authenticated)
	internalSchemes := normalizeScopeList(spec.AuthSchemes.InternalOnly)
	for _, name := range append(append([]string(nil), authenticatedSchemes...), internalSchemes...) {
		if _, ok := schemes[name]; !ok {
			return nil, nil, nil, fmt.Errorf("apptheory: secure openapi auth scheme %s is not defined", name)
		}
	}
	return schemes, authenticatedSchemes, internalSchemes, nil
}

func secureOpenAPIJoins(registered map[string]SecureRoute, described map[string]OpenAPIRouteSpec, order, authenticatedSchemes, internalSchemes []string) ([]secureOpenAPIJoin, []OpenAPIRouteSpec, error) {
	joins := make([]secureOpenAPIJoin, 0, len(order))
	emitted := map[string]struct{}{}
	legacyRoutes := make([]OpenAPIRouteSpec, 0, len(order))
	for _, key := range order {
		route := registered[key]
		description := described[key]
		proxy, proxyName, emittedPath := secureProxyPath(description.Path)
		emittedKey := strings.ToUpper(description.Method) + " " + emittedPath
		if _, exists := emitted[emittedKey]; exists {
			return nil, nil, fmt.Errorf("apptheory: secure openapi emitted route %s collides", emittedKey)
		}
		emitted[emittedKey] = struct{}{}
		if proxy {
			description.Path = emittedPath
			description.Request.Fields = secureEnsureProxyField(description.Request.Fields, proxyName)
		}
		if route.Posture == AuthPostureOptional || route.Posture == AuthPostureAuthenticated || route.Posture == AuthPostureAuthenticatedAnyOf {
			if len(authenticatedSchemes) == 0 {
				return nil, nil, errors.New("apptheory: secure openapi authenticated scheme binding is required")
			}
		}
		if route.Posture == AuthPostureInternalOnly && len(internalSchemes) == 0 {
			return nil, nil, errors.New("apptheory: secure openapi internal scheme binding is required")
		}
		joins = append(joins, secureOpenAPIJoin{route: route, description: description, emittedPath: emittedPath, proxy: proxy, proxyName: proxyName})
		legacyRoutes = append(legacyRoutes, description)
	}
	return joins, legacyRoutes, nil
}

func secureDecorateOpenAPI(doc map[string]any, schemes map[string]any, joins []secureOpenAPIJoin, authenticatedSchemes, internalSchemes []string) error {
	components, ok := doc["components"].(map[string]any)
	if !ok {
		return errors.New("apptheory: secure openapi components invariant")
	}
	components["securitySchemes"] = schemes
	doc["x-apptheory-contract-mode"] = "secure-v1"

	paths, ok := doc["paths"].(map[string]any)
	if !ok {
		return errors.New("apptheory: secure openapi paths invariant")
	}
	for _, join := range joins {
		pathItem, pathOK := paths[join.emittedPath].(map[string]any)
		if !pathOK {
			return errors.New("apptheory: secure openapi path invariant")
		}
		operation, operationOK := pathItem[strings.ToLower(join.description.Method)].(map[string]any)
		if !operationOK {
			return errors.New("apptheory: secure openapi operation invariant")
		}
		operation["x-apptheory-auth-posture"] = string(join.route.Posture)
		if len(join.route.Scopes) > 0 {
			operation["x-apptheory-required-scopes"] = append([]string(nil), join.route.Scopes...)
		}
		if join.proxy {
			operation["x-apptheory-proxy"] = true
		}
		operation["security"] = secureOperationSecurity(join.route, authenticatedSchemes, internalSchemes)
	}
	return nil
}

// GenerateOpenAPIJSON returns the byte-stable canonical JSON encoding of secure OpenAPI generation.
func (a *SecureApp) GenerateOpenAPIJSON(spec SecureOpenAPISpec) ([]byte, error) {
	doc, err := a.GenerateOpenAPI(spec)
	if err != nil {
		return nil, err
	}
	var buf bytes.Buffer
	encoder := json.NewEncoder(&buf)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(doc); err != nil {
		return nil, err
	}
	return bytes.TrimSuffix(buf.Bytes(), []byte("\n")), nil
}

func secureOperationSecurity(route SecureRoute, authenticated, internal []string) []any {
	out := []any{}
	appendSchemes := func(names []string, scopes []string) {
		for _, name := range names {
			out = append(out, map[string]any{name: append([]string{}, scopes...)})
		}
	}
	switch route.Posture {
	case AuthPosturePublic:
		return out
	case AuthPostureOptional:
		appendSchemes(authenticated, nil)
		return append(out, map[string]any{})
	case AuthPostureAuthenticated:
		appendSchemes(authenticated, route.Scopes)
	case AuthPostureAuthenticatedAnyOf:
		for _, name := range authenticated {
			for _, scope := range route.Scopes {
				out = append(out, map[string]any{name: []string{scope}})
			}
		}
	case AuthPostureInternalOnly:
		appendSchemes(internal, nil)
	}
	return out
}

func secureProxyPath(path string) (bool, string, string) {
	segments := splitPath(path)
	if len(segments) == 0 {
		return false, "", "/"
	}
	last := strings.TrimSpace(segments[len(segments)-1])
	if !strings.HasPrefix(last, "{") || !strings.HasSuffix(last, "+}") {
		return false, "", path
	}
	name := strings.TrimSpace(last[1 : len(last)-2])
	segments[len(segments)-1] = "{" + name + "}"
	return true, name, "/" + strings.Join(segments, "/")
}

func secureEnsureProxyField(fields []OpenAPIFieldSpec, name string) []OpenAPIFieldSpec {
	out := append([]OpenAPIFieldSpec(nil), fields...)
	for i := range out {
		if strings.EqualFold(strings.TrimSpace(out[i].Source), bindSourcePath) && strings.TrimSpace(out[i].Name) == name {
			out[i].Required = true
			return out
		}
	}
	return append(out, OpenAPIFieldSpec{Field: name, Source: bindSourcePath, Name: name, Type: openAPITypeString, Required: true})
}

func secureCanonicalSecuritySchemes(input map[string]map[string]any) (map[string]any, error) {
	out := make(map[string]any, len(input))
	seen := map[uintptr]bool{}
	for rawName, scheme := range input {
		name := strings.TrimSpace(rawName)
		if name == "" {
			return nil, errors.New("apptheory: secure openapi security scheme name is required")
		}
		if _, exists := out[name]; exists {
			return nil, fmt.Errorf("apptheory: secure openapi security scheme %s is duplicated", name)
		}
		value, err := secureCanonicalJSONValue(reflect.ValueOf(scheme), seen)
		if err != nil {
			return nil, fmt.Errorf("apptheory: secure openapi security scheme %s: %w", name, err)
		}
		out[name] = value
	}
	return out, nil
}

func secureCanonicalJSONValue(value reflect.Value, seen map[uintptr]bool) (any, error) {
	if !value.IsValid() {
		return nil, nil
	}
	if value.Kind() == reflect.Interface {
		if value.IsNil() {
			return nil, nil
		}
		return secureCanonicalJSONValue(value.Elem(), seen)
	}
	switch value.Kind() {
	case reflect.Bool:
		return value.Bool(), nil
	case reflect.String:
		return value.String(), nil
	case reflect.Map:
		return secureCanonicalJSONObject(value, seen)
	case reflect.Slice, reflect.Array:
		return secureCanonicalJSONArray(value, seen)
	default:
		return nil, fmt.Errorf("value type %s is not allowed", value.Kind())
	}
}

func secureCanonicalJSONObject(value reflect.Value, seen map[uintptr]bool) (map[string]any, error) {
	if value.Type().Key().Kind() != reflect.String {
		return nil, errors.New("object keys must be strings")
	}
	ptr := value.Pointer()
	if ptr != 0 && seen[ptr] {
		return nil, errors.New("cyclic value is not allowed")
	}
	if ptr != 0 {
		seen[ptr] = true
		defer delete(seen, ptr)
	}
	out := map[string]any{}
	iter := value.MapRange()
	for iter.Next() {
		item, err := secureCanonicalJSONValue(iter.Value(), seen)
		if err != nil {
			return nil, err
		}
		out[iter.Key().String()] = item
	}
	return out, nil
}

func secureCanonicalJSONArray(value reflect.Value, seen map[uintptr]bool) ([]any, error) {
	ptr := uintptr(0)
	if value.Kind() == reflect.Slice && !value.IsNil() {
		ptr = value.Pointer()
	}
	if ptr != 0 && seen[ptr] {
		return nil, errors.New("cyclic value is not allowed")
	}
	if ptr != 0 {
		seen[ptr] = true
		defer delete(seen, ptr)
	}
	out := make([]any, value.Len())
	for i := 0; i < value.Len(); i++ {
		item, err := secureCanonicalJSONValue(value.Index(i), seen)
		if err != nil {
			return nil, err
		}
		out[i] = item
	}
	return out, nil
}

File: ts/src/secure-openapi.ts
import type { SecureRoute } from "./app.js";
import {
  generateOpenAPI,
  type OpenAPIDocument,
  type OpenAPIFieldSpec,
  type OpenAPIRouteSpec,
} from "./openapi.js";

/** Document-level secure posture scheme bindings. */
export interface OpenAPIAuthSchemes {
  authenticated: string[];
  internalOnly: string[];
}

/** Exact HTTP description accepted by SecureApp OpenAPI generation. */
export interface SecureOpenAPISpec {
  title: string;
  version: string;
  routes: readonly OpenAPIRouteSpec[];
  securitySchemes: Record<string, Record<string, unknown>>;
  authSchemes: OpenAPIAuthSchemes;
}

function canonicalRouteKey(methodValue: string, pathValue: string) {
  const method = String(methodValue ?? "")
    .trim()
    .toUpperCase();
  if (!method) throw new Error("apptheory: route method is empty");
  let path = String(pathValue ?? "").trim();
  const queryIndex = path.indexOf("?");
  if (queryIndex >= 0) path = path.slice(0, queryIndex);
  path = path.trim();
  if (!path) path = "/";
  if (!path.startsWith("/")) path = `/${path}`;
  const rawSegments = path === "/" ? [] : path.slice(1).split("/");
  const canonical: string[] = [];
  for (let index = 0; index < rawSegments.length; index += 1) {
    let segment = String(rawSegments[index] ?? "").trim();
    if (!segment) throw new Error("apptheory: invalid route pattern");
    if (segment.startsWith(":") && segment.length > 1) {
      segment = `{${segment.slice(1)}}`;
    }
    if (segment.startsWith("{") && segment.endsWith("}")) {
      let name = segment.slice(1, -1).trim();
      const proxy = name.endsWith("+");
      if (proxy) name = name.slice(0, -1).trim();
      if (!name || name.includes("{") || name.includes("}")) {
        throw new Error("apptheory: invalid route pattern");
      }
      if (proxy && index !== rawSegments.length - 1) {
        throw new Error("apptheory: invalid route pattern");
      }
      canonical.push(`{${name}${proxy ? "+" : ""}}`);
      continue;
    }
    if (segment.includes("{") || segment.includes("}")) {
      throw new Error("apptheory: invalid route pattern");
    }
    canonical.push(segment);
  }
  path = canonical.length > 0 ? `/${canonical.join("/")}` : "/";
  return { key: `${method} ${path}`, method, path };
}

function normalizeNames(values: readonly string[] | undefined): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const raw of values ?? []) {
    const value = String(raw ?? "").trim();
    if (!value || seen.has(value)) continue;
    seen.add(value);
    out.push(value);
  }
  return out;
}

function copyJSONValue(value: unknown, seen: WeakSet<object>): unknown {
  if (
    value === null ||
    typeof value === "string" ||
    typeof value === "boolean"
  ) {
    return value;
  }
  if (typeof value === "number" || typeof value === "undefined") {
    throw new Error("numeric and undefined values are not allowed");
  }
  if (typeof value !== "object") {
    throw new Error(`value type ${typeof value} is not allowed`);
  }
  if (seen.has(value)) throw new Error("cyclic value is not allowed");
  seen.add(value);
  try {
    if (Array.isArray(value)) {
      return value.map((item) => copyJSONValue(item, seen));
    }
    const prototype = Object.getPrototypeOf(value) as unknown;
    if (prototype !== Object.prototype && prototype !== null) {
      throw new Error("runtime-specific objects are not allowed");
    }
    const out: Record<string, unknown> = {};
    for (const key of Reflect.ownKeys(value)) {
      if (typeof key !== "string") {
        throw new Error("object keys must be strings");
      }
      out[key] = copyJSONValue((value as Record<string, unknown>)[key], seen);
    }
    return out;
  } finally {
    seen.delete(value);
  }
}

function normalizeSecuritySchemes(
  input: Record<string, Record<string, unknown>>,
): Record<string, unknown> {
  const copied = copyJSONValue(input ?? {}, new WeakSet()) as Record<
    string,
    unknown
  >;
  const out: Record<string, unknown> = {};
  for (const [rawName, scheme] of Object.entries(copied)) {
    const name = rawName.trim();
    if (!name) {
      throw new Error(
        "apptheory: secure openapi security scheme name is required",
      );
    }
    if (Object.hasOwn(out, name)) {
      throw new Error(
        `apptheory: secure openapi security scheme ${name} is duplicated`,
      );
    }
    out[name] = scheme;
  }
  return out;
}

function proxyPath(path: string): {
  proxy: boolean;
  name: string;
  path: string;
} {
  const segments = path === "/" ? [] : path.slice(1).split("/");
  const last = segments.at(-1) ?? "";
  if (!last.startsWith("{") || !last.endsWith("+}")) {
    return { proxy: false, name: "", path };
  }
  const name = last.slice(1, -2).trim();
  segments[segments.length - 1] = `{${name}}`;
  return { proxy: true, name, path: `/${segments.join("/")}` };
}

function proxyFields(
  fields: readonly OpenAPIFieldSpec[] | undefined,
  name: string,
): OpenAPIFieldSpec[] {
  const out = [...(fields ?? [])].map((field) => ({ ...field }));
  const existing = out.find(
    (field) => field.source === "path" && String(field.name).trim() === name,
  );
  if (existing) {
    existing.required = true;
    return out;
  }
  out.push({
    field: name,
    source: "path",
    name,
    type: "string",
    required: true,
  });
  return out;
}

function operationSecurity(
  route: SecureRoute,
  authenticated: string[],
  internal: string[],
): unknown[] {
  if (route.posture === "public") return [];
  const names = route.posture === "internal_only" ? internal : authenticated;
  if (route.posture === "authenticated_any_of") {
    return names.flatMap((name) =>
      (route.scopes ?? []).map((scope) => ({ [name]: [scope] })),
    );
  }
  const scopes =
    route.posture === "authenticated" ? [...(route.scopes ?? [])] : [];
  const out: unknown[] = names.map((name) => ({ [name]: scopes }));
  if (route.posture === "optional") out.push({});
  return out;
}

export function generateSecureOpenAPI(
  routes: readonly SecureRoute[],
  spec: SecureOpenAPISpec,
): OpenAPIDocument {
  const registered = new Map<string, SecureRoute>();
  for (const route of routes) {
    if (route.surface !== "http") continue;
    registered.set(canonicalRouteKey(route.method, route.path).key, route);
  }
  const described = new Map<string, OpenAPIRouteSpec>();
  const order: string[] = [];
  for (const input of spec.routes) {
    const canonical = canonicalRouteKey(input.method, input.path);
    if (described.has(canonical.key)) {
      throw new Error(
        `apptheory: secure openapi route ${canonical.key} is duplicated`,
      );
    }
    described.set(canonical.key, {
      ...input,
      method: canonical.method,
      path: canonical.path,
    });
    order.push(canonical.key);
  }
  for (const key of registered.keys()) {
    if (!described.has(key))
      throw new Error(`apptheory: secure openapi missing route ${key}`);
  }
  for (const key of described.keys()) {
    if (!registered.has(key))
      throw new Error(`apptheory: secure openapi extra route ${key}`);
  }

  const schemes = normalizeSecuritySchemes(spec.securitySchemes ?? {});
  const authenticated = normalizeNames(spec.authSchemes?.authenticated);
  const internal = normalizeNames(spec.authSchemes?.internalOnly);
  for (const name of [...authenticated, ...internal]) {
    if (!(name in schemes)) {
      throw new Error(
        `apptheory: secure openapi auth scheme ${name} is not defined`,
      );
    }
  }

  const joins: Array<{
    route: SecureRoute;
    description: OpenAPIRouteSpec;
    emittedPath: string;
    proxy: boolean;
  }> = [];
  const emitted = new Set<string>();
  for (const key of order) {
    const route = registered.get(key) as SecureRoute;
    let description = described.get(key) as OpenAPIRouteSpec;
    const proxy = proxyPath(description.path);
    const emittedKey = `${description.method.toUpperCase()} ${proxy.path}`;
    if (emitted.has(emittedKey))
      throw new Error(
        `apptheory: secure openapi emitted route ${emittedKey} collides`,
      );
    emitted.add(emittedKey);
    if (
      (route.posture === "optional" ||
        route.posture === "authenticated" ||
        route.posture === "authenticated_any_of") &&
      authenticated.length === 0
    ) {
      throw new Error(
        "apptheory: secure openapi authenticated scheme binding is required",
      );
    }
    if (route.posture === "internal_only" && internal.length === 0) {
      throw new Error(
        "apptheory: secure openapi internal scheme binding is required",
      );
    }
    if (proxy.proxy) {
      description = {
        ...description,
        path: proxy.path,
        request: {
          ...description.request,
          fields: proxyFields(description.request?.fields, proxy.name),
        },
      };
    }
    joins.push({
      route,
      description,
      emittedPath: proxy.path,
      proxy: proxy.proxy,
    });
  }

  const document = generateOpenAPI({
    title: spec.title,
    version: spec.version,
    routes: joins.map((join) => join.description),
  }) as unknown as Record<string, unknown>;
  const components = document["components"] as Record<string, unknown>;
  components["securitySchemes"] = schemes;
  document["x-apptheory-contract-mode"] = "secure-v1";
  const paths = document["paths"] as Record<
    string,
    Record<string, Record<string, unknown>>
  >;
  for (const join of joins) {
    const operation =
      paths[join.emittedPath]?.[join.description.method.toLowerCase()];
    if (!operation)
      throw new Error("apptheory: secure openapi operation invariant");
    operation["x-apptheory-auth-posture"] = join.route.posture;
    if ((join.route.scopes?.length ?? 0) > 0) {
      operation["x-apptheory-required-scopes"] = [...(join.route.scopes ?? [])];
    }
    if (join.proxy) operation["x-apptheory-proxy"] = true;
    operation["security"] = operationSecurity(
      join.route,
      authenticated,
      internal,
    );
  }
  return document as OpenAPIDocument;
}

function compareUnicodeScalars(left: string, right: string): number {
  const a = Array.from(left, (ch) => ch.codePointAt(0) ?? 0);
  const b = Array.from(right, (ch) => ch.codePointAt(0) ?? 0);
  for (let i = 0; i < Math.min(a.length, b.length); i += 1) {
    if (a[i] !== b[i]) return (a[i] ?? 0) - (b[i] ?? 0);
  }
  return a.length - b.length;
}

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(",")}]`;
  const record = value as Record<string, unknown>;
  return `{${Object.keys(record)
    .sort(compareUnicodeScalars)
    .map((key) => `${JSON.stringify(key)}:${stableStringify(record[key])}`)
    .join(",")}}`;
}

export function generateSecureOpenAPIJSON(
  routes: readonly SecureRoute[],
  spec: SecureOpenAPISpec,
): string {
  return stableStringify(generateSecureOpenAPI(routes, spec));
}
